core/src/main/java/feast/core/model/FeatureSet.java

```diff
@@ -94,19 +61,40 @@
   private long maxAgeSeconds;
 
   // Entity fields inside this feature set
-  @ElementCollection(fetch = FetchType.EAGER)
-  @CollectionTable(name = "entities", joinColumns = @JoinColumn(name = "feature_set_id"))
-  @Fetch(FetchMode.SUBSELECT)
-  private Set<Field> entities;
+  @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
+  @JoinTable(
+      name = "feature_set_entities",
+      joinColumns = @JoinColumn(name = "feature_set_id"),
+      inverseJoinColumns = {
+        @JoinColumn(name = "entities_name"),
+        @JoinColumn(name = "entities_project"),
+        @JoinColumn(name = "entities_feature_set_id"),
+        @JoinColumn(name = "entities_version")
+      },
+      indexes = {
+        @Index(
+            name = "idx_jobs_feature_set_entities_feature_set_id",
+            columnList = "feature_set_id"),
+      })
+  private Set<Entity> entities;
 
   // Feature fields inside this feature set
-  @ElementCollection(fetch = FetchType.EAGER)
-  @CollectionTable(
-      name = "features",
+  @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
+  @JoinTable(
+      name = "feature_set_features",
       joinColumns = @JoinColumn(name = "feature_set_id"),
-      uniqueConstraints = @UniqueConstraint(columnNames = {"name", "project", "version"}))
-  @Fetch(FetchMode.SUBSELECT)
-  private Set<Field> features;
+      inverseJoinColumns = {
+        @JoinColumn(name = "features_name"),
+        @JoinColumn(name = "features_project"),
+        @JoinColumn(name = "features_feature_set_id"),
+        @JoinColumn(name = "features_version")
+      },
+      indexes = {
+        @Index(
+            name = "idx_jobs_feature_set_features_feature_set_id",
+            columnList = "feature_set_id"),
+      })
```

Comment:
I really don't want to delay this PR any more because we do have some time pressure, but since its a data model I think its worth making sure that we are getting it right. In previous discussions ([Add computation and retrieval PR](https://github.com/gojek/feast/pull/612#issuecomment-616520831)) we agreed on the removal of Fields in a kind of evolution of the existing schema.

However, looking at the schema now it still seems kind of complicated (needing join tables). I am not sure if it is necessarily complicated or unnecessarily. I don't know all the Springisms and how that affects our data model, but just going by how this could be modeled in SQL without Spring. Is there anything that prevents us from following the more normalized approach as follows?

```
Project
- id (PK)
- name
- ...

Feature set
- id (PK)
- project_id (unique 1/3 )
- name (unique 1/3 )
- version (unique 1/3 )
- ...

Feature (or Entity)
- id (PK)
- feature_set_id (unique 1/2 )
- name (unique 1/2 )
```

All of the existing functionality like feature references would still hold, I am thinking purely from a persistence standpoint.

Reply:
Having a placeholder id means that whenever you want to join something to that object you will have to look up the actual object in the database to get the pkey instead of letting the orm do the heavy lifting. (case in point, how this PR joins job <> feature set).

Reply:
> (case in point, how this PR joins job <> feature set).

Could you elaborate a bit more with an example and counterexample? It's hard to extrapolate from code what ORMs will do with a bunch of annotations, or their entirely implicit behavior, without running the code oneself with SQL logging turned up.

Reply:
With a proper id (containing the actual identifying elements of an object, like `FeatureReference` in this PR), you are able to reference that object to perform joins on or mutate the object using those identifying elements. This is useful because we get these references from either the protos or user requests.

If feature set had an id, [this](https://github.com/gojek/feast/blob/9ba8fee75a3b5976bb819a1bee8ab5f7fd9c484e/core/src/main/java/feast/core/service/JobCoordinatorService.java#L204) doesn't need to be done, because i could just instantiate the FeatureSet from the proto and then make the status changes, and the correct FeatureSet would be updated.

In the absence of such an id, the placeholder id has to be retrieved each time to point to that exact object whenever you want to join something else to it.

Reply:
I'm sorry if what i said is a little incoherent :( the model proposed by willem is possible, and there are ways around what i've mentioned, but having natural ids would allow us to move between protos and db objects more naturally when forming relationships between feast objects is what I mean

Reply:
But i'm not an expert, so if you guys think it's the better approach to use placeholder ids then i'd be happy to implement it

Reply:
I'm happy to figure out the pros/cons of the natural/surrogate key approach, but do we agree the model above makes sense?

I mean specifically in the sense of not having join tables between features/entities and feature sets, and not persisting parent fields like `project` or `feature set` in the children (entities, features).

Reply:
> elaborate a bit more with an example and counterexample

Yea I think this would be super useful. I understand the reasoning above, but I am trying to find out what this actually looks like in practice.